File: tools/event_prep/us_twt_merge_2010_2023.py
"""Merge the two US-Twt configs into one continuous 2010-2023 site.

Decided in the 2026-08-29 review call: the split existed only because the input
data differs either side of 2017, and "you don't want to split it up simply
because the input data is wrong or is inconsistent. We want to have consistent
input data for the model."

Tillage is the inconsistency. The 2010-2016 events carry 14 tillage events from
Knox; the 2017-2023 monitoring product carries none, which is a gap in the product
rather than a change in practice. Per the call, the literature is primary and
monitoring is the fallback, so the documented Knox pattern is extended forward:

    disking      planting - 21 d,  tillage_eff 0.5
    ring rolling planting -  7 d,  tillage_eff 0.2

That interval is exact in all seven Knox years, not fitted.
"""
import json, datetime, copy

CAL = "/home/aritra/Downloads/projects/ccmmf/calibration"
D = datetime.date.fromisoformat

a = json.load(open(f"{CAL}/runs/us_twt_2010/events.json"))   # 2010-2016, literature
b = json.load(open(f"{CAL}/runs/us_twt/events.json"))        # 2017-2023, monitoring

assert a["site_id"] == b["site_id"] == "US-Twt"

# --- verify the tillage rule holds in the literature years before extending it ---
lit_till = [e for e in a["events"] if e["event_type"] == "tillage"]
lit_plant = {e["date"][:4]: D(e["date"]) for e in a["events"] if e["event_type"] == "planting"}
offsets = {}
for e in lit_till:
    yr = e["date"][:4]
    if yr in lit_plant:
        offsets.setdefault(e["implement"], set()).add((lit_plant[yr] - D(e["date"])).days)
for imp, offs in sorted(offsets.items()):
    print(f"  {imp}: planting minus {sorted(offs)} d")
    assert len(offs) == 1, f"{imp} offset is not constant: {sorted(offs)}"
DISK_D, ROLL_D = offsets["disking"].pop(), offsets["ring rolling"].pop()

# --- backfill tillage for the monitoring years from that rule ---
new_till = []
for e in sorted(b["events"], key=lambda x: x["date"]):
    if e["event_type"] != "planting":
        continue
    p = D(e["date"])
    for imp, off, eff in (("disking", DISK_D, 0.5), ("ring rolling", ROLL_D, 0.2)):
        new_till.append({
            "event_type": "tillage",
            "date": (p - datetime.timedelta(days=off)).isoformat(),
            "tillage_eff_0to1": eff,
            "implement": imp,
            "provenance_class": "assumed",
            "source": "calval:managements/delta-tillage-0001",
            "provenance_note": (
                f"The monitoring product records no tillage for this parcel, which is a gap in the "
                f"product rather than a change in practice. Knox documents {imp} every year at "
                f"planting minus {off} d with tillage_eff {eff}, exact in all seven of 2010-2016, so "
                f"that pattern is carried forward onto the monitoring planting date of {e['date']}."
            ),
        })
print(f"  backfilled tillage events for 2017-2023: {len(new_till)}")

merged = copy.deepcopy(b)
merged["pecan_events_version"] = "0.1.1"
merged["provenance"] = {
    "parcel_id": 590073,
    "crop": "R1",
    "generator": "merge of the 2010-2016 literature build and the 2017-2023 monitoring build",
    "note": (
        "One continuous 2010-2023 series. 2010-2016 management is derived from Knox et al. 2016 "
        "with flood and drain dates from the AmeriFlux WTD series; 2017-2023 comes from the "
        "statewide monitoring product, which does not cover this parcel before 2017. Tillage for "
        "2017-2023 is carried forward from the Knox pattern because the product records none. "
        "Per-event source and provenance_class record which applies to each event."
    ),
}
merged["events"] = sorted(a["events"] + b["events"] + new_till,
                          key=lambda e: (e["date"], e["event_type"]))

out = f"{CAL}/runs/us_twt/events.json"
json.dump(merged, open(out, "w"), indent=1)

import collections
ds = [e["date"] for e in merged["events"]]
print(f"  merged: {len(merged['events'])} events  {ds[0]} -> {ds[-1]}")
print(f"  by type: {dict(collections.Counter(e['event_type'] for e in merged['events']))}")
till_yr = collections.Counter(e["date"][:4] for e in merged["events"] if e["event_type"]=="tillage")
print(f"  tillage per year: {dict(sorted(till_yr.items()))}")

File: tools/event_prep/us_twt_merge_2010_2023.R
#!/usr/bin/env Rscript
# Merge the 2010-2016 literature build and the 2017-2023 monitoring build into one
# series, backfilling tillage by carrying forward the Knox planting-offset rule.

suppressMessages(library(jsonlite))

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3L)
  stop("usage: us_twt_merge_2010_2023.R <literature.json> <monitoring.json> <out.json>")

a <- jsonlite::fromJSON(args[1], simplifyDataFrame = FALSE)  # 2010-2016, literature
b <- jsonlite::fromJSON(args[2], simplifyDataFrame = FALSE)  # 2017-2023, monitoring
stopifnot(identical(a$site_id, b$site_id))

etype <- function(ev, t) Filter(function(e) identical(e$event_type, t), ev)

# verify the tillage rule holds in the literature years before extending it
lit_till <- etype(a$events, "tillage")
lit_plant <- setNames(
  lapply(etype(a$events, "planting"), function(e) as.Date(e$date)),
  vapply(etype(a$events, "planting"), function(e) substr(e$date, 1, 4), character(1)))

offsets <- list()
for (e in lit_till) {
  yr <- substr(e$date, 1, 4)
  if (!is.null(lit_plant[[yr]]))
    offsets[[e$implement]] <- union(offsets[[e$implement]],
                                    as.integer(lit_plant[[yr]] - as.Date(e$date)))
}
for (imp in sort(names(offsets))) {
  cat(sprintf("  %s: planting minus %s d\n", imp, paste(sort(offsets[[imp]]), collapse = ", ")))
  if (length(offsets[[imp]]) != 1L)
    stop(sprintf("%s offset is not constant: %s", imp, paste(sort(offsets[[imp]]), collapse = ", ")))
}
DISK_D <- offsets[["disking"]]; ROLL_D <- offsets[["ring rolling"]]

# backfill tillage for the monitoring years from that rule
new_till <- list()
plantings <- etype(b$events, "planting")
plantings <- plantings[order(vapply(plantings, function(e) e$date, character(1)))]
for (e in plantings) {
  p <- as.Date(e$date)
  for (k in seq_len(2L)) {
    imp <- c("disking", "ring rolling")[k]
    off <- c(DISK_D, ROLL_D)[k]
    eff <- c(0.5, 0.2)[k]
    new_till[[length(new_till) + 1L]] <- list(
      event_type = "tillage",
      date = format(p - off, "%Y-%m-%d"),
      tillage_eff_0to1 = eff,
      implement = imp,
      provenance_class = "assumed",
      source = "calval:managements/delta-tillage-0001",
      provenance_note = sprintf(paste(
        "The monitoring product records no tillage for this parcel, which is a gap in the",
        "product rather than a change in practice. Knox documents %s every year at planting",
        "minus %d d with tillage_eff %s, exact in all seven of 2010-2016, so that pattern is",
        "carried forward onto the monitoring planting date of %s."), imp, off, eff, e$date))
  }
}
cat(sprintf("  backfilled tillage events for 2017-2023: %d\n", length(new_till)))

merged <- b
merged$pecan_events_version <- "0.1.1"
merged$provenance <- list(
  parcel_id = 590073L, crop = "R1",
  generator = "merge of the 2010-2016 literature build and the 2017-2023 monitoring build",
  note = paste(
    "One continuous 2010-2023 series. 2010-2016 management is derived from Knox et al. 2016",
    "with flood and drain dates from the AmeriFlux WTD series; 2017-2023 comes from the",
    "statewide monitoring product, which does not cover this parcel before 2017. Tillage for",
    "2017-2023 is carried forward from the Knox pattern because the product records none.",
    "Per-event source and provenance_class record which applies to each event."))

all_ev <- c(a$events, b$events, new_till)
ord <- order(vapply(all_ev, function(e) e$date, character(1)),
             vapply(all_ev, function(e) e$event_type, character(1)))
merged$events <- all_ev[ord]
write(jsonlite::toJSON(merged, auto_unbox = TRUE, pretty = 1, digits = NA), args[3])

ds <- vapply(merged$events, function(e) e$date, character(1))
cat(sprintf("  merged: %d events  %s -> %s\n", length(merged$events), ds[1], ds[length(ds)]))
print(table(vapply(merged$events, function(e) e$event_type, character(1))))
